File: blog/burak/DanglemayrO2full/psectgs.m
function ps=psectgs(xhatp, nhat, filename, direction)

load gramschmidt.mat;
load timeev.mat;

%Plane condition:

%direction=-1; %Now a function input

U = @(x) dot((x-xhatp),nhat)*direction;

n = size(xhatGS,2);
tol= 1e-6; %Tolerance for 'zero'

ps = [xhatp;0]; %Assign a dummy value to ps vector to be able to fill afterwards

for i = 1:n-1
	%If plane equation change sign from negative to positive, look for the intersection:
	if U(xhatGS(:,i)) < 0 && U(xhatGS(:,i+1)) > 0 
		%Integration until the plane condition is satisfied:
			
		xdummyGS = xhatGS(:,i);
		xdummyfull = x(:,i);

		DotProduct = U(xdummyGS); 
		deltatadaptive = deltat/2;
		pstime = (i-1)*deltat;
		j=1	;
		while abs(DotProduct) > tol
			j=j+1;
			%Adaptive integration: 
			
			xint=integrator(xdummyfull, deltatadaptive, deltatadaptive);
			
			
			pstime = pstime + deltatadaptive;
			
			xdummyfull = xint(:,2);
			
			xdummyhat = findxhat(xdummyfull,[1;1;0;0]);
			xdummyhat = xdummyhat(1:size(xdummyhat)-1);
			
			xdummyGS = findGS(xdummyhat);

			DotProduct = U(xdummyGS);
			
			% If the Poincare section is passed get back in integration
			
			if DotProduct > 0 % If the Poincare section is passed
				
				xdummyfull = xint(:,1);
				pstime = pstime - deltatadaptive;
				deltatadaptive = deltatadaptive/2;
				
				%xdummyhat = findxhat(xdummyfull,[1;1;0;0])
				%xdummyhat = xdummyhat(1:size(xdummyhat)-1)
				%xdummy = findGS(xdummyhat)
				%DotProduct = U(xdummy)
				
			end
		
			if j>100
				fprintf('not converging, exiting...\n')
				break;
			end
		end
		
		ps(:,size(ps,2)+1) = [xdummyGS;
							  pstime];
		%ps(:,size(ps,2)+1) = [xdummyGS;
							  %pstime];
	end
	
end

ps = ps(:, 2:size(ps,2));
save(filename, 'ps','xhatp')

File: blog/burak/PorterKnobloch/python/psectgs.py
#
#psectgs.py
#
import numpy as np
from scipy.integrate import odeint
import twomode
import onslicesolver

#Load parameters:
pars = np.loadtxt('data/parameters.dat')

xhatpsect = np.array([0,0,0], float)
nhat = np.array([0,1,0], float)
direction = 1 # set 1 or -1 to choose between directions of piercing the Poincare section
tolerance = 1e-9

#Poincare hyperplane equation:
def U(x):
	cond=np.dot(x-xhatpsect, nhat) * direction
	return cond

# ODE solver parameters:
abserr = 1.0e-12
relerr = 1.0e-9

def computeps(xhatGS, p=pars):
	
	deltat = xhatGS[1,0]-xhatGS[0,0] # Read deltat from the data
	
	#dummy assignment for ps:
	ps = np.array([[0, 0, 0, 0], [0,0,0,0]], float)
	
	#Look for intersections:
	for i in range(len(xhatGS)-1):
		if U(xhatGS[i, 1:4])<0 and U(xhatGS[i+1, 1:4])>0:
			xdummyGS = xhatGS[i, 1:4]
			deltatadaptive = deltat/2
			pstime = xhatGS[i,0]
			j=1
			condition = U(xdummyGS)
			xhatphidummy=np.concatenate((twomode.gramschmidt2ssp(xdummyGS), np.array([0],float)))
			while np.abs(condition)>tolerance:
				j=j+1
				#Adaptive integration:
				t = [float(0), deltatadaptive]
				xhatsol = onslicesolver.integrate(xhatphidummy, p, t, abserr, relerr)
				xhatphidummy=xhatsol[1, :]
				pstime = pstime + deltatadaptive
				xdummyGS = twomode.ssp2gramschmidt(xhatphidummy[0:4])
				condition = U(xdummyGS)
				
				if condition > 0:
					xhatphidummy=xhatsol[0,:]
					pstime = pstime - deltatadaptive
					deltatadaptive = deltatadaptive/2
				
				if j > 100:
					print 'not converging, exiting...'
					break
			pstime = np.array([pstime], float)
			psdummy = np.concatenate((pstime, xdummyGS))
		
			ps = np.append(ps, [psdummy],0)
	
	ps = ps[2:len(ps),:]

	return ps

#If the module is called as a script, compute the Poincare section from the
#data and print:

if __name__ == "__main__":
	
	xhatGS = np.loadtxt('data/gramschmidt.dat', unpack=True) #Gram-Schmidt data
	xhatGS=xhatGS.transpose() # xhatGS = [t, x1hatGS, y1hatGS, y2hatGS]
	
	ps=computeps(xhatGS)
	
	#Print the poincare section
	for pss in ps:
		print pss[0], pss[1], pss[2], pss[3]
